vision_agent/agent/agent_coder.py

```diff
@@ -0,0 +1,170 @@
+
+
+def write_tests(question: str, code: str, model: LLM) -> str:
+    prompt = TEST.format(
+        question=question,
+        code=code,
+    )
+    completion = model(prompt)
+    return preprocess_data(completion)
+
+
+def preprocess_data(code: str) -> str:
+    if "```python" in code:
+        code = code[code.find("```python") + len("```python") :]
+        code = code[: code.find("```")]
+    return code
+
+
+def parse_file_name(s: str) -> str:
+    # We only output png files
+    return "".join([p for p in s.split(" ") if p.endswith(".png")])
+
+
+def write_program(question: str, feedback: str, model: LLM) -> str:
+    prompt = PROGRAM.format(
+        docstring=TOOLS_DOCSTRING, question=question, feedback=feedback
+    )
+    completion = model(prompt)
+    return preprocess_data(completion)
+
+
+def write_debug(question: str, code: str, feedback: str, model: LLM) -> str:
+    prompt = DEBUG.format(
+        docstring=UTILITIES_DOCSTRING,
+        code=code,
+        question=question,
+        feedback=feedback,
+    )
+    completion = model(prompt)
+    return preprocess_data(completion)
+
+
+def execute_tests(code: str, tests: str) -> Dict[str, Union[str, bool]]:
+    full_code = f"{IMPORT_HELPER}\n{code}\n{tests}"
+    return check_correctness(full_code, 20.0)
+
+
+def run_visual_tests(
+    question: str, code: str, viz_file: str, feedback: str, model: LMM
+) -> Dict[str, Union[str, bool]]:
+    prompt = VISUAL_TEST.format(
+        docstring=TOOLS_DOCSTRING,
+        code=code,
+        question=question,
+        feedback=feedback,
+    )
+    completion = model(prompt, images=[viz_file])
+    # type is from the prompt
+    return json.loads(completion)  # type: ignore
+
+
+def fix_bugs(code: str, tests: str, result: str, feedback: str, model: LLM) -> str:
+    prompt = FIX_BUG.format(completion=code, test_case=tests, result=result)
+    completion = model(prompt)
+    return preprocess_data(completion)
+
+
+class AgentCoder(Agent):
+    """AgentCoder is based off of the AgentCoder paper https://arxiv.org/abs/2312.13010
+    and it's open source code https://github.com/huangd1999/AgentCoder with some key
+    differences. AgentCoder comprises of 3 components: a coder agent, a tester agent,
+    and an executor. The tester agents writes code to test the code written by the coder
+    agent, but in our case because we are solving a vision task it's difficult to write
+    testing code. We instead have the tester agent write code to visualize the output
+    of the code written by the coder agent. If the code fails, we pass it back to the
+    coder agent to fix the bug, if it succeeds we pass it to a visual tester agent, which
+    is an LMM model like GPT4V, to visually inspect the output and make sure it looks
+    good."""
+
+    def __init__(
+        self,
+        coder_agent: Optional[LLM] = None,
+        tester_agent: Optional[LLM] = None,
+        visual_tester_agent: Optional[LMM] = None,
```

Comment:
So the visual_tester_agent acts like reflection ?

Reply:
Yeah this is very similar to vision agent, where the reflection agent is an LMM that looks at the visualized output.